internal/envoy/auth/jwt_filter.go

```diff
@@ -0,0 +1,242 @@
+				Match: &envoy_config_route_v3.RouteMatch{
+					PathSpecifier: &envoy_config_route_v3.RouteMatch_Path{
+						Path: path,
+					},
+				},
+				RequirementType: &envoy_jwt_v3.RequirementRule_Requires{
+					Requires: &envoy_jwt_v3.JwtRequirement{
+						RequiresType: &envoy_jwt_v3.JwtRequirement_ProviderAndAudiences{
+							ProviderAndAudiences: &envoy_jwt_v3.ProviderWithAudiences{
+								ProviderName: provider.Name,
+								Audiences:    provider.Audiences,
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+
+	if err := jwtConfig.ValidateAll(); err != nil {
+		return nil, err
+	}
+
+	return anypb.New(&jwtConfig)
+}
+
+func DNSNameClusterName(cluster *DNSNameCluster) string {
+	return strings.Join([]string{"dnsname", cluster.Scheme, cluster.Address}, "/")
+}
+
+// DNSNameCluster is a cluster that routes directly to a DNS
+// name (i.e. not a Kubernetes service).
+type DNSNameCluster struct {
+	Address            string
+	Scheme             string
+	Port               int
+	DNSLookupFamily    string
+	UpstreamValidation *PeerValidationContext
+}
+
+// PeerValidationContext defines how to validate the certificate on the upstream service.
+type PeerValidationContext struct {
+	// CACertificate holds a reference to the Secret containing the CA to be used to
+	// verify the upstream connection.
+	CACertificate *Secret
+	// SubjectName holds an optional subject name which Envoy will check against the
+	// certificate presented by the upstream.
+	SubjectName string
+	// SkipClientCertValidation when set to true will ensure Envoy requests but
+	// does not verify peer certificates.
+	SkipClientCertValidation bool
+	// ForwardClientCertificate adds the selected data from the passed client TLS certificate
+	// to the x-forwarded-client-cert header.
+	ForwardClientCertificate *ClientCertificateDetails
+	// CRL holds a reference to the Secret containing the Certificate Revocation List.
+	// It is used to check for revocation of the peer certificate.
+	CRL *Secret
+	// OnlyVerifyLeafCertCrl when set to true, only the certificate at the end of the
+	// certificate chain will be subject to validation by CRL.
+	OnlyVerifyLeafCertCrl bool
+	// OptionalClientCertificate when set to true will ensure Envoy does not require
+	// that the client sends a certificate but if one is sent it will process it.
+	OptionalClientCertificate bool
+}
+
+// ClientCertificateDetails defines which parts of the client certificate will be forwarded.
+type ClientCertificateDetails struct {
+	// Subject of the client cert.
+	Subject bool
+	// Client cert in URL encoded PEM format.
+	Cert bool
+	// Client cert chain (including the leaf cert) in URL encoded PEM format.
+	Chain bool
+	// DNS type Subject Alternative Names of the client cert.
+	DNS bool
+	// URI type Subject Alternative Name of the client cert.
+	URI bool
+}
+
+// Secret represents a K8s Secret for TLS usage as a DAG Vertex. A Secret is
+// a leaf in the DAG.
+type Secret struct {
```

Comment:
Does this belong in the `auth` package?

Reply:
Good catch.

I've removed unused code.